### ISSUE TITLE
Deleting containers VMs in vSphere may cause containers that use the same base image to fail to create.

### ISSUE THREAD

**Issue by mlh78750** [vmware/vic#2928]
See #2927. Users should not delete container vms from the vsphere side. However, this is extremely non-intuitive for those users. 

So we need to protect the VCH from having the images removed out from under them by VC. We should have a test that covers these conditions and that is tracked in #2925

**Comment by mlh78750** [comment]
Possible fix from @hickeng: I'm only aware of one workaround, which is to create an additional child VMDK of all the entries in the disk chain so that VC will consider the disks still in use.

**Comment by fdawg4l** [comment]
Weirdly, I'm having trouble reproducing this.  I know im doing something wrong but working on what that is.  :-)

```
$ docker -H 192.168.8.174:2376 --tls ps -a                                                                                                                
CONTAINER ID        IMAGE               COMMAND             CREATED             STATUS              PORTS               NAMES
4f20f21b111c        busybox             "/bin/top"          41 seconds ago      Up 39 seconds                           tender_heisenberg

$ govc ls vm
/ha-datacenter/vm/virtual-container-host
/ha-datacenter/vm/tender_heisenbe-4f20f21b111c53134ce3302ccfa3627c548e9d5478c2af1a31334c237e049701

$ govc vm.destroy /ha-datacenter/vm/tender_heisenbe-4f20f21b111c53134ce3302ccfa3627c548e9d5478c2af1a31334c237e049701

$ govc ls vm
/ha-datacenter/vm/virtual-container-host

$ docker -H 192.168.8.174:2376 --tls ps -a
CONTAINER ID        IMAGE               COMMAND             CREATED             STATUS              PORTS               NAMES

$ docker -H 192.168.8.174:2376 --tls images
REPOSITORY          TAG                 IMAGE ID            CREATED             SIZE
busybox             latest              e292aa76ad3b        2 weeks ago         1.093 MB

$ docker -H 192.168.8.174:2376 --tls run -d busybox /bin/top                                                                                              
1af6d08f64b4ec703156348ff3d7f7275041c4667a5b603e74ea0d2ae3cbbb29

$ docker -H 192.168.8.174:2376 --tls ps -a                                                                                                                
CONTAINER ID        IMAGE               COMMAND             CREATED             STATUS              PORTS               NAMES
1af6d08f64b4        busybox             "/bin/top"          18 seconds ago      Up 15 seconds                           trusting_chandrasekhar
```

**Comment by fdawg4l** [comment]
Ah.  Per @hickeng 

> yup - VC only. ESX doesn't technically support linked clones (which is what disk chains enable)

OK.  I'll try again with a VC.

**Comment by dougm** [comment]
Note that an additional child VMDK does not prevent the disk chain from being destroyed, unless said child is attached to VM.  ESX does support linked clones, but unlike VC, ESX doesn't delete the parent disks when the disks are no longer referenced by VMs.  To reproduce on VC, using a VCH with no other container VMs:

```sh
% docker pull busybox

% id=$(docker create busybox date)

% govc vm.destroy *-$id

% docker create busybox date
Error response from daemon: Server error from portlayer: Invalid configuration for device '1'.
```

(I ran into this stuff working on the vSAN leak)

**Comment by stuclem** [comment]
Proposed text for the 0.9 release notes:

----

- <b>Deleting container VMs by using the vSphere Client can remove the underlying image. [#1234](https://github.com/vmware/vic/issues/1234)</b><br>If you delete a container VM by using the vSphere Client, attempts to create other containers that use the same base image containers can fail if the base image has been removed.
  
    Workaround: As stated in the [documentation](https://vmware.github.io/vic-product/assets/files/html/0.8/vic_admin/interop.html), always use Docker commands to perform operations on containers. Do not use the vSphere Client to perform operations on container VMs.
---

@mlh78750 does this cover it?

**Comment by mlh78750** [comment]
@stuclem I think this covers it from a release note perspective. I still want us to figure out how to fix this overall. This is a huge usability issue with the main value that we deliver with the product.

**Comment by stuclem** [comment]
Thanks @mlh78750. Removing kind/note.

**Comment by hickeng** [comment]
The recent FCD change is the only thing that allows us to address this cleanly - even then it will instead mean that VM folder and read/write disk are left behind after a VM deletion so we'd need to then perform cleanup of the orphaned disks.

Without FCD we could disable the deletion method from the UI, but that introduces it's own issues.
I really don't have a good answer to this.

**Comment by hickeng** [comment]
Spoken to Dani/GSS and his recommendation is to disable the delete method and KB how to do it without disrupting the image hierarchy if an out-of-band deletion is required.

**Comment by fdawg4l** [comment]
> The recent FCD change is the only thing that allows us to address this cleanly - even then it will instead mean that VM folder and read/write disk are left behind after a VM deletion so we'd need to then perform cleanup of the orphaned disks.

By recent change, I'm assuming you mean in the vsphere API? 

> his recommendation is to disable the delete method 

Which delete method are we disabling and where @hickeng?

Also, prio was changed to high.  Should the fix in this case be to document in the kb not to do the deletion out of band and if required, how to do it?

**Comment by fdawg4l** [comment]
Clarified with @hickeng during sprint planning-  Fix is to add the vm deletion method to the black list in vpshere for the vapp to prevent the user from being able to delete the container out of band.

**Comment by dougm** [comment]
Spoke with @chengwang86 in slack, @hickeng was suggesting to use the ManagedEntity.DisabledMethods property.  However, afaict, there is no way to modify that field from the API.  Search opengrok for `GetDisabledMethod`, you'll find the implementation are either a static list or generated based on runtime properties/state.

**Comment by chengwang86** [comment]
Update:

As a workaround for OOB deletion of containerVMs (if not avoidable):
1. detach the hard disk from the containerVM
2. Delete the containerVM out-of-band
3. login to the esxi host where the containerVM was on via ssh and delete the datastore folder of this containerVM manually.

**Comment by chengwang86** [comment]
There are several ways to disable VM deletion from vsphere web client:

**Method 1:** 
- How to: Use `AuthorizationManager.disableMethods()` to disable the `Destroy_Task` from vsphere UI.
See http://www.virtuallyghetto.com/2016/07/how-to-easily-disable-vmotion-cross-vcenter-vmotion-for-a-particular-virtual-machine.html
- Behavior: If this is done, the `delete vm from disk` option from the right-click menu of the containerVM will be `grey` (disabled).
- Downside: However, this requires the `Global.disableMethods` privilege, which may not be available for an `ops-user` with limited privileges. 

**Method 2:**
- How to: Set the `managedBy` property in `VirtualMachineConfigSpec`. See https://pubs.vmware.com/vsphere-65/index.jsp#com.vmware.wssdk.smssdk.doc/vim.vm.ConfigSpec.html#field_detail
- Behavior: If you try to delete or reconfigure a VM with this property set, the user will see a confirmation message box containing the extension name that is managing the VM, e.g., `this entity is managed by the vSphere Integrated Container Engine. It is not recommended to modify it directly. Instead use the VIC commands if you want to make changes.`
- Downside: The requirement for using this property is for your service to be registered as VcExtension using `ExtensionManager` here http://pubs.vmware.com/vsphere-65/topic/com.vmware.vsphere.ext_solutions.doc/GUID-185B96FB-B84B-4301-B6E1-8F2FA4A67A2A.html

**Comment by chengwang86** [comment]
Update:

It seems like the scratch image will be deleted when the user performs OOB VM deletion as long as there is no other VM that references it.

I have tried two scenarios: (I have ubuntu and busybox images pulled before testing)
- Case A: VM_1 with ubuntu and VM_2 with busybox. Deleting VM_2 OOB does not cause any problem.
- Case B: only a single VM with busybox. Deleting that VM causes the issue discussed here. Starting a VM with ubuntu also fails.

**Comment by chengwang86** [comment]
Based on discussion with @fdawg4l , we might be able to user WORM or extended attributes to prevent the base image from being deleted by OOB deletion of VMs if these are supported on vmfs. This would rely on relevant vmomi api to be exposed.

**Comment by fdawg4l** [comment]
@chengwang86 all of that is very unlikely.  But, I'll look into it.  The options are just workarounds for a gap in our API usage and the user's intent.

**Comment by dougm** [comment]
To clarify on Method 2, extensions are not deprecated, but LoginByExtension is (client cert auth).  It is possible to create an extension just to configure the ManagedBy field.  I tried it out and the biggest problem is it just adds more text to a dialog box that is likely not read anyhow.  The delete confirmation dialog box looks exactly the same, just more words.

I also tried Method 1 and confirmed it greys out the delete option but does not prevent API method calls to Destroy_Task.

**Comment by dougm** [comment]
Even if/when we find a way to avoid the parent images from being removed, we should still disable VM delete from the UI since we can never prevent data volumes from being deleted if the VM is deleted from the UI.

I've opened vmware/govmomi#713 , the integration with VIC will be straight forward.  Regarding the Global.DisableMethods role, if the VIC user does not have the permission, we can just log a message rather than fail.

**Comment by dougm** [comment]
Note that while First Class Disks are included in the 6.5 vmodl, the flag to prevent their deletion in this case is *not*.  It is in vmkernel-main, but we won't be able to make use of it in VIC until the next vSphere release.
